### PR TITLE
fix: schema cache invalidation

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- #5064

## What's changed and what's your intention?
Adds cache registry to datanode, fix the issue that cache invalidation cannot propogate the datanode cache.

All tests in #5063 are tested.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/cache/src/lib.rs [hunk 1]
```diff
@@ -37,9 +37,40 @@ pub const TABLE_INFO_CACHE_NAME: &str = "table_info_cache";
...
+        kv_backend.clone(),
+    ));
+
+    CacheRegistryBuilder::default()
+        .add_cache(table_info_cache)
+        .add_cache(schema_cache)
+        .build()
+}
+
 pub fn build_fundamental_cache_registry(kv_backend: KvBackendRef) -> CacheRegistry {
```
Comment (killme2008) [inline src/cache/src/lib.rs:74]: Can we add a doc for this function too? Declare its usage.

Comment (killme2008) [reply src/cache/src/lib.rs:74]: Is it only used frontend? If so, maybe we can rename it into `build_frontend_cache_registry`.

Comment (v0y4g3r) [reply src/cache/src/lib.rs:83]: > Is it only used frontend? If so, maybe we can rename it into `build_frontend_cache_registry`.

It's shared by frontend and flownode. I added some doc in 737f5184d0

**Review by WenyXu — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@fengjiachun @waynexia please take a look, #5063 is waiting for this PR.

File: src/cache/src/lib.rs [hunk 2]
```diff
@@ -95,12 +135,24 @@ pub fn build_fundamental_cache_registry(kv_backend: KvBackendRef) -> CacheRegist
         kv_backend.clone(),
     ));
 
+    // Builds schema cache
+    let cache = CacheBuilder::new(DEFAULT_CACHE_MAX_CAPACITY)
+        .time_to_live(DEFAULT_CACHE_TTL)
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/cache/src/lib.rs:138]: fixed in https://github.com/GreptimeTeam/greptimedb/compare/737f5184d094a52674e7702136e7c3d059313618..5b08d74e4018ce81f6686f98b5eba5116b60961f

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/cache/table/table_schema.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    _cache: &'a Cache<TableId, Arc<SchemaName>>,
+    _ident: &'a CacheIdent,
+) -> BoxFuture<'a, error::Result<()>> {
+    Box::pin(std::future::ready(Ok(())))
+}
+
+/// Never invalidates table id schema cache.
+fn filter(_ident: &CacheIdent) -> bool {
+    false
+}
```
Comment (WenyXu) [inline src/common/meta/src/cache/table/table_schema.rs:76]: It's better to invalidate cache while dropping the table

File: src/cache/src/lib.rs [hunk 4]
```diff
@@ -95,12 +133,30 @@ pub fn build_fundamental_cache_registry(kv_backend: KvBackendRef) -> CacheRegist
...
+        .time_to_idle(DEFAULT_CACHE_TTI)
+        .build();
+    let schema_cache = Arc::new(new_schema_cache(
+        SCHEMA_CACHE_NAME.to_string(),
+        cache,
+        kv_backend.clone(),
+    ));
+
+    let table_schema_name_cache = Arc::new(new_table_schema_cache(
```
Comment (WenyXu) [inline src/cache/src/lib.rs:147]:
```suggestion
    let table_id_schema_cache = Arc::new(new_table_id_schema_cache(
```

File: src/cache/src/lib.rs [hunk 5]
```diff
@@ -95,12 +133,30 @@ pub fn build_fundamental_cache_registry(kv_backend: KvBackendRef) -> CacheRegist
...
+        kv_backend,
+    ));
     CacheRegistryBuilder::default()
         .add_cache(table_info_cache)
         .add_cache(table_name_cache)
         .add_cache(table_route_cache)
         .add_cache(view_info_cache)
         .add_cache(table_flownode_set_cache)
+        .add_cache(schema_cache)
+        .add_cache(table_schema_name_cache)
```
Comment (WenyXu) [inline src/cache/src/lib.rs:159]:
```suggestion
        .add_cache(table_id_schema_cache)
```

**Review by WenyXu — APPROVED** [review]
Rest LGTM

File: src/cache/src/lib.rs [hunk 6]
```diff
@@ -95,12 +133,30 @@ pub fn build_fundamental_cache_registry(kv_backend: KvBackendRef) -> CacheRegist
...
+        .time_to_live(DEFAULT_CACHE_TTL)
+        .time_to_idle(DEFAULT_CACHE_TTI)
+        .build();
+    let schema_cache = Arc::new(new_schema_cache(
+        SCHEMA_CACHE_NAME.to_string(),
+        cache,
+        kv_backend.clone(),
+    ));
+
+    let table_id_schema_cache = Arc::new(new_table_id_schema_cache(
```
Comment (killme2008) [inline src/cache/src/lib.rs:147]:
```suggestion
    let table_id_schema_cache = Arc::new(new_table_schema_cache(
```

**Pull request merged**